lib/vector-core/src/fanout.rs

```diff
@@ -63,165 +62,330 @@ impl Fanout {
         self.sinks.push((id, Some(sink)));
     }
 
+    /// Remove an existing sink as an output.
     fn remove(&mut self, id: &ComponentKey) {
-        let i = self.sinks.iter().position(|(n, _)| n == id);
-        let i = i.expect("Didn't find output in fanout");
-
-        let (_id, removed) = self.sinks.remove(i);
+        if let Some(i) = self.sinks.iter().position(|(n, _)| n == id) {
+            let (_id, _removed) = self.sinks.remove(i);
+        }
+    }
 
-        if let Some(mut removed) = removed {
-            tokio::spawn(async move { removed.close().await });
+    /// Replace an existing sink as an output.
+    ///
+    /// If the `sink` passed is `None`, operation of the `Fanout` will be paused until a `Some`
+    /// with the same key is received. This allows for cases where the previous version of
+    /// a stateful sink must be dropped before the new version can be created.
+    fn replace(&mut self, id: ComponentKey, sink: Option<BufferSender<Event>>) {
+        if let Some((_, existing)) = self.sinks.iter_mut().find(|(n, _)| n == &id) {
+            *existing = sink;
+        } else {
+            // send might have been in-flight at the time, in which case it will be cancelled
+            // panic!("Tried to replace a sink that's not already present");
```

Comment:
Intentionally left?

Reply:
Nope, but similar to the comment above in terms of semantics.